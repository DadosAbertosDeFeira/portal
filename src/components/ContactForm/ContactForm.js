/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { toast } from 'react-toastify';
import { useForm } from 'react-hook-form';
import Input from '../form/Input/Input';
import Textarea from '../form/Textarea/Textarea';
import validations from '../../utils/validations/validations';

const ContactForm = () => {
  const form = useForm({
    mode: 'onSubmit',
    shouldUseNativeValidation: true,
    defaultValues: {
      name: '',
      email: '',
      message: '',
    },
  });

  const onSubmit = async (data) => {
    const response = fetch('/', {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: new URLSearchParams({
        'form-name': 'contact',
        ...data,
      }).toString(),
    }).then((r) => {
      form.reset();
      return r;
    });

    await toast.promise(response, {
      pending: 'Enviando mensagem',
      error: 'Ocorreu um erro ao enviar sua mensagem',
      success: 'Sua mensagem foi enviada com sucesso',
    });
  };

  return (
    <form
      className="flex flex-col w-full gap-y-4"
      onSubmit={form.handleSubmit(onSubmit)}
    >
      <h1 className="text-2xl">Fale conosco</h1>
      <Input
        name="name"
        label="Nome"
        control={form.control}
        rules={{ required: 'Esse campo é obrigatório.' }}
      />
      <Input
        name="email"
        label="Email"
        control={form.control}
        rules={{ validate: validations.get('email') }}
      />
      <Textarea
        name="message"
        label="Mensagem"
        control={form.control}
        rows={3}
        rules={{ required: 'Esse campo é obrigatório.' }}
      />
      <button type="submit" className="button-outline">
        Enviar
      </button>
    </form>
  );
};

export default ContactForm;
