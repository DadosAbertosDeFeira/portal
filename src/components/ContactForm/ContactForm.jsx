/* eslint-disable import/extensions */
import React from "react";
import { useForm } from "react-hook-form";
import { toast } from "react-toastify";
import { StyledButton } from "../molecules/StyledButton";
import { InputForm } from "../molecules/InputForm";
import { TextareaForm } from "../molecules/TextareaForm";

import validations from "../../utils/validations/validations";

const ContactForm = () => {
  const form = useForm({
    mode: "onSubmit",
    shouldUseNativeValidation: true,
    defaultValues: {
      name: "",
      email: "",
      message: "",
    },
  });

  const onSubmit = async (data) => {
    const response = fetch("/", {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        "form-name": "contact",
        ...data,
      }).toString(),
    }).then((r) => {
      form.reset();
      return r;
    });

    await toast.promise(response, {
      pending: "Enviando mensagem",
      error: "Ocorreu um erro ao enviar sua mensagem",
      success: "Sua mensagem foi enviada com sucesso",
    });
  };

  return (
    <form
      className="flex w-full flex-col gap-y-4"
      onSubmit={form.handleSubmit(onSubmit)}
    >
      <h1 className="text-2xl">Fale conosco</h1>
      <InputForm
        name="name"
        label="Nome"
        control={form.control}
        rules={{ required: "Esse campo é obrigatório." }}
      />
      <InputForm
        name="email"
        label="Email"
        control={form.control}
        rules={{ validate: validations.get("email") }}
      />
      <TextareaForm
        name="message"
        label="Mensagem"
        control={form.control}
        rows={3}
        rules={{ required: "Esse campo é obrigatório." }}
      />
      <StyledButton type="submit" variant="outline">
        Enviar
      </StyledButton>
    </form>
  );
};

export default ContactForm;
