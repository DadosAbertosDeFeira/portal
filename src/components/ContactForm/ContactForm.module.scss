@import '../../styles/variables.scss';

.title {
  font-size: 1.5rem;
}

.contactForm {
  display: flex;
  flex-direction: column;
  max-width: 400px;
  width: 100%;
}

.contactFormInput,
.contactFormTextarea {
  width: 100%;
  border: 1px solid $color-gray-2;
  border-radius: 4px;
  padding-left: 1rem;
}

.contactFormInput {
  height: 40px;
}

.contactFormInput + .contactFormInput {
  margin-top: 1rem;
}

.contactFormTextarea {
  margin-top: 1rem;
  padding-top: calc(20px - 0.5rem);
}

.contactFormSubmitButton {
  margin-top: 1rem;
  height: 40px;
  width: 100%;
}
