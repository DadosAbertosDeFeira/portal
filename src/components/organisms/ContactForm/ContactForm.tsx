import { validations } from "@utils/validations";
import { Button } from "atoms/Button";
import { Input } from "atoms/Input";
import { Textarea } from "atoms/Textarea";
import type { SubmitHandler } from "react-hook-form";
import { Controller, useForm } from "react-hook-form";
import { toast } from "react-toastify";

export interface ContactFormModel {
  name: string;
  email: string;
  message: string;
}

export function ContactForm() {
  const form = useForm<ContactFormModel>({
    mode: "onSubmit",
    shouldUseNativeValidation: true,
    defaultValues: {
      name: "",
      email: "",
      message: "",
    },
  });

  const onSubmit: SubmitHandler<ContactFormModel> = async (data) => {
    const toastId = toast.loading("Enviando mensagem");

    try {
      const response = await fetch("/", {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: new URLSearchParams({
          "form-name": "contact",
          ...data,
        }).toString(),
      });
      if (response.status !== 200) throw new Error();

      toast.update(toastId, {
        render: "Sua mensagem foi enviada com sucesso",
        type: "success",
        isLoading: false,
        autoClose: 5000,
      });

      form.reset();
    } catch (e) {
      toast.update(toastId, {
        render: "Ocorreu um erro ao enviar sua mensagem",
        type: "error",
        isLoading: false,
        autoClose: 5000,
      });
    }
  };

  return (
    <form
      className="flex w-full flex-col gap-y-4"
      onSubmit={form.handleSubmit(onSubmit)}
    >
      <h1 className="text-2xl">Fale conosco</h1>
      <Controller
        name="name"
        control={form.control}
        rules={{ required: "Esse campo é obrigatório" }}
        render={({ field, fieldState: { error } }) => (
          <Input
            errorText={error?.message}
            label="Nome"
            variant="outline"
            hideLabel
            filled
            {...field}
          />
        )}
      />

      <Controller
        name="email"
        control={form.control}
        rules={{ validate: validations.get("email") }}
        render={({ field, fieldState: { error } }) => (
          <Input
            errorText={error?.message}
            label="Email"
            variant="outline"
            hideLabel
            filled
            {...field}
          />
        )}
      />

      <Controller
        name="message"
        control={form.control}
        rules={{ required: "Esse campo é obrigátorio" }}
        render={({ field, fieldState: { error } }) => (
          <Textarea
            errorText={error?.message}
            label="Mensagem"
            hideLabel
            {...field}
          />
        )}
      />

      <Button type="submit" variant="outline">
        Enviar
      </Button>
    </form>
  );
}
