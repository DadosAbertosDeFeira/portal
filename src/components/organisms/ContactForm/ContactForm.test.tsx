import { fireEvent, render, screen, waitFor } from "@testing-library/react";

import { ContactForm } from "./ContactForm";

jest.mock("react-toastify", () => ({
  __esModule: true,
  toast: {
    loading: jest.fn(),
    update: jest.fn(),
  },
}));

describe("ContactForm", () => {
  const fetch = jest.fn().mockResolvedValue({});

  beforeEach(() => {
    global.fetch = fetch;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const makeSut = ({ ...props }: Partial<{}>) => {
    return render(<ContactForm {...props} />);
  };

  it("renders correctly", () => {
    const { asFragment } = makeSut({});

    expect(asFragment).toMatchSnapshot();
  });

  it("should submit form", async () => {
    const mockForm = {
      "form-name": "contact",
      name: "name",
      email: "name@domain.com",
      message: "Exercitation adipisicing sunt est cupidatat enim",
    };

    makeSut({});

    const name = await screen.findByLabelText("Nome");
    fireEvent.input(name, { target: { value: mockForm.name } });

    const email = await screen.findByLabelText("Email");
    fireEvent.input(email, { target: { value: mockForm.email } });

    const message = await screen.findByLabelText("Mensagem");
    fireEvent.input(message, { target: { value: mockForm.message } });

    const button = await screen.findByRole("button");

    fireEvent.click(button);

    await waitFor(() => {
      expect(fetch).toHaveBeenCalledWith("/", {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: new URLSearchParams(mockForm).toString(),
      });
    });
  });
});
