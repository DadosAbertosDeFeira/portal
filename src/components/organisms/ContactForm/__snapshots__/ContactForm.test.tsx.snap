// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm renders correctly 1`] = `[Function]`;
